wagtailmedia/views/chooser.py

```diff
@@ -105,3 +115,45 @@ def media_chosen(request, media_id):
         None,
         json_data={'step': 'media_chosen', 'result': get_media_json(media)}
     )
+
+
+@permission_checker.require('add')
+def chooser_upload(request, media_type):
+    Media = get_media_model()
+    MediaForm = get_media_form(Media)
+
+    if request.method == 'POST':
+        media = Media(uploaded_by_user=request.user, type=media_type)
+        form = MediaForm(
+            request.POST, request.FILES, instance=media, user=request.user,
+            prefix='media-chooser-upload'
+        )
+        if form.is_valid():
+            form.save()
+
+            # Reindex the media entry to make sure all tags are indexed
+            for backend in get_search_backends():
+                backend.add(media)
+
+            return render_modal_workflow(
+                request, None, None, None,
+                json_data={'step': 'media_chosen',
+                           'result': get_media_json(media)}
+            )
+    else:
+        media = Media(uploaded_by_user=request.user, type=media_type)
+        form = MediaForm(
+            instance=media, user=request.user, prefix='media-chooser-upload')
+
+    media_files = Media.objects.order_by('-created_at')
+    paginator, media_files = paginate(request, media_files, per_page=10)
+
+    context = {
+        'media_files': media_files,
+        'uploadform': form,
+        'media_type': media_type,
```

Comment:
Since chooser_upload is only called as a POST what do you think about setting missing keys as hardcoded like is_searching = False?
Well we wont be able to get possible query parameters since they are lost at POST. Or should we preserve them?

Reply:
The bit above inside the `else:` is not needed, but this code here is what happens on a POST request when the form is not valid. So we need to have this set up with real values for all of the context. For `is_searching` - yes that will always be False, but the rest should match what’s defined in the main `chooser` view.